File: wisp/infra/extensions.py
"""ExtensionHost — unified extension system.

Replaces: plugin_registry.py, hooks.py, mcp.py, skills.py
with one lifecycle-managed host.

All extensions implement:
  - start() → None
  - stop() → None
  - tools() → list[dict]  (OpenAI-style function schemas)
  - intercept(event: dict) → {"action": "allow"|"block", "reason": str}
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


class ExtensionHost:
    """One system for all extensions."""

    def __init__(self):
        self._extensions: list[Any] = []

    def register(self, ext: Any) -> None:
        """Register an extension and start it."""
        try:
            ext.start()
            self._extensions.append(ext)
            logger.debug("Extension started: %s", getattr(ext, "name", type(ext).__name__))
        except Exception as exc:
            logger.warning("Extension %s failed to start: %s", getattr(ext, "name", "?"), exc)
            # Do NOT append broken extensions

    def tools(self) -> list[dict]:
        """Aggregate tools from all extensions."""
        tools: list[dict] = []
        for ext in self._extensions:
            try:
                ext_tools = ext.tools()
                if ext_tools:
                    tools.extend(ext_tools)
            except Exception as exc:
                logger.warning(
                    "Extension %s tools() failed: %s",
                    getattr(ext, "name", type(ext).__name__),
                    exc,
                )
        return tools

    def intercept(self, event: dict) -> dict:
        """Run event through all extensions. First block wins.

        If an extension raises an exception, the call is denied (fail-closed).
        """
        for ext in self._extensions:
            try:
                result = ext.intercept(event)
                if result.get("action") == "block":
                    return result
            except Exception as exc:
                logger.exception(
                    "Extension %s intercept() failed: %s — denying by default",
                    getattr(ext, "name", type(ext).__name__),
                    exc,
                )
                return {"action": "block", "reason": f"Extension error: {exc}"}
        return {"action": "allow"}

    def start(self) -> None:
        """Extensions already started during register() — kept for ServiceRegistry contract."""

    def stop(self) -> None:
        """Stop all extensions in reverse registration order."""
        for ext in reversed(self._extensions):
            try:
                ext.stop()
                logger.debug("Extension stopped: %s", getattr(ext, "name", type(ext).__name__))
            except Exception as exc:
                logger.warning(
                    "Extension %s stop() failed: %s",
                    getattr(ext, "name", "?"),
                    exc,
                )
        self._extensions.clear()

File: wisp-ts/src/extensions.ts
/** ExtensionHost — unified extension system with lifecycle management. */

export interface Extension {
  name: string;
  start(): void;
  stop(): void;
  tools(): Array<Record<string, unknown>>;
  intercept(event: Record<string, unknown>): { action: "allow" | "block"; reason?: string };
}

export class ExtensionHost {
  private _extensions: Extension[] = [];

  register(ext: Extension): void {
    try {
      ext.start();
      this._extensions.push(ext);
    } catch (exc) {
      console.warn(`Extension ${ext.name} failed to start: ${exc}`);
    }
  }

  tools(): Array<Record<string, unknown>> {
    const tools: Array<Record<string, unknown>> = [];
    for (const ext of this._extensions) {
      try {
        const extTools = ext.tools();
        if (extTools?.length) tools.push(...extTools);
      } catch (exc) {
        console.warn(`Extension ${ext.name} tools() failed: ${exc}`);
      }
    }
    return tools;
  }

  intercept(event: Record<string, unknown>): { action: "allow" | "block"; reason?: string } {
    for (const ext of this._extensions) {
      try {
        const result = ext.intercept(event);
        if (result.action === "block") return result;
      } catch (exc) {
        console.warn(`Extension ${ext.name} intercept() failed: ${exc} — denying by default`);
        return { action: "block", reason: `Extension error: ${exc}` };
      }
    }
    return { action: "allow" };
  }

  start(): void {
    // Extensions already started during register()
  }

  stop(): void {
    for (const ext of [...this._extensions].reverse()) {
      try {
        ext.stop();
      } catch (exc) {
        console.warn(`Extension ${ext.name} stop() failed: ${exc}`);
      }
    }
    this._extensions.length = 0;
  }
}

/** PluginExtension — wraps a simple tool provider as an extension. */
export class PluginExtension implements Extension {
  name: string;
  private _tools: Array<Record<string, unknown>>;

  constructor(name: string, tools: Array<Record<string, unknown>>) {
    this.name = name;
    this._tools = tools;
  }

  start(): void {}
  stop(): void {}

  tools(): Array<Record<string, unknown>> {
    return this._tools;
  }

  intercept(_event: Record<string, unknown>): { action: "allow" | "block" } {
    return { action: "allow" };
  }
}

/** SkillExtension — exposes workspace skills as prefixed tools. */
export class SkillExtension implements Extension {
  name = "skills";
  private _workspace: string;

  constructor(workspace: string) {
    this._workspace = workspace;
  }

  start(): void {}
  stop(): void {}

  tools(): Array<Record<string, unknown>> {
    // In a real implementation, would discover skills from workspace
    return [];
  }

  intercept(_event: Record<string, unknown>): { action: "allow" | "block" } {
    return { action: "allow" };
  }
}
